pkg/scheduler/queue/query_component_capacity.go

```diff
@@ -0,0 +1,179 @@
+	Ingester     QueryComponent = "ingester"
+)
+
+// cannot import constants from frontend/v2 due to import cycle
+// these are attached to the request's AdditionalQueueDimensions by the frontend.
+const ingesterQueueDimension = "ingester"
+const storeGatewayQueueDimension = "store-gateway"
+const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
+
+// QueryComponentNameForRequest parses the expected query component from annotations by the frontend.
+func QueryComponentNameForRequest(req *SchedulerRequest) string {
+	var expectedQueryComponent string
+	if len(req.AdditionalQueueDimensions) > 0 {
+		expectedQueryComponent = req.AdditionalQueueDimensions[0]
+	}
+
+	return expectedQueryComponent
+}
+
+// queryComponentFlags interprets annotations by the frontend for the expected query component,
+// and flags whether a query request is expected to be served by the ingesters, store-gateways, or both.
+func queryComponentFlags(queryComponentName string) (isIngester, isStoreGateway bool) {
+	// Annotations from the frontend representing "both" or "unknown" do not need to be checked explicitly.
+	// Conservatively, we assume a query request will be served by both query components
+	// as we prefer to overestimate query component capacity utilization rather than underestimate.
+	isIngester, isStoreGateway = true, true
+
+	switch queryComponentName {
+	case ingesterQueueDimension:
+		isStoreGateway = false
+	case storeGatewayQueueDimension:
+		isIngester = false
+	}
+	return isIngester, isStoreGateway
+}
+
+// QueryComponentCapacity tracks requests from the time they are forwarded to a querier
+// to the time are completed by the querier or failed due to cancel, timeout, or disconnect.
+// Unlike the Scheduler's schedulerInflightRequests, tracking begins only when the request is sent to a querier.
+//
+// Scheduler-Querier inflight requests are broken out by the query component flags,
+// representing whether the query request will be served by the ingesters, store-gateways, or both.
+// Query requests utilizing both ingesters and store-gateways are tracked in the atomics for both component,
+// therefore the sum of inflight requests by component is likely to exceed the inflight requests total.
+type QueryComponentCapacity struct {
+	// targetReservedCapacity sets the portion of querier-worker connections we attempt to reserve
+	// for queries to the less-loaded query component when the query queue becomes backlogged.
+	targetReservedCapacity float64
+
+	ingesterInflightRequests     *atomic.Int64
+	storeGatewayInflightRequests *atomic.Int64
+	querierInflightRequestsTotal *atomic.Int64
+
+	querierInflightRequestsGauge *prometheus.GaugeVec
+}
+
+// DefaultReservedQueryComponentCapacity reserves 1 / 3 of querier-worker connections
+// for the query component utilizing fewer of the available connections.
+const DefaultReservedQueryComponentCapacity = 0.33
+
+const MinReservedQueryComponentCapacity = 0.1
```

Comment:
Honestly I did not have a good reason for 10% other than it didn't seem all that useful to reserve much less than this
If the point is to keep queries for one component moving when the other component is struggling, it seems like a waste to allow the struggling component to dominate the capacity. This would cap it at 90% at least.

We could allow it to be down to zero, which would be like disabling the feature. That could make sense if we plan to make this a configurable feature in the future

Reply:
I think having a way to disable this would be valuable, especially while we're still experimenting with it.

Reply:
Set to not have a minumum